aten/src/TH/generic/THTensorEvenMoreMath.cpp

```diff
@@ -29,17 +29,37 @@ void THTensor_(zero)(THTensor *r_)
 
 void THTensor_(maskedFill)(THTensor *tensor, THByteTensor *mask, scalar_t value)
 {
-  TH_TENSOR_APPLY2(scalar_t, tensor, unsigned char, mask,
-                   if (*mask_data > 1)
-                   {
-                     THFree(mask_counter);
-                     THFree(tensor_counter);
-                     THError("Mask tensor can take 0 and 1 values only");
-                   }
-                   else if (*mask_data == 1)
-                   {
-                     *tensor_data = value;
-                   });
+  int serial_path = 0;
+#ifdef _OPENMP
+  int inOMP = omp_in_parallel();
+  if (inOMP) {
+    serial_path = 1;
+  } else {
+    int64_t tensor_size = THTensor_(nElement)(tensor);
+    int tensor_contig = THTensor_(isContiguous)(tensor);
+    int mask_contig = THTensor_(isContiguous)(mask);
+    TH_TENSOR_APPLY2_OMP(tensor_size, tensor_contig, mask_contig,
+      scalar_t, tensor, unsigned char, mask,
+      if (*mask_data > 1) {
+        THError("Mask tensor can take 0 and 1 values only");
+      } else if (*mask_data == 1) {
+        *tensor_data = value;
+      },
+      TH_OMP_OVERHEAD_THRESHOLD);
+  }
+#else
+  serial_path = 1;
+#endif
```

Comment:
Is the `serial_path` variable really needed? It seems to me that this function could have easily been structured like this:
```cpp
#ifdef _OPENMP
if (!omp_in_parallel()) {
  ...
  TH_TENSOR_APPLY2_OMP(...)
  return;
}
#endif
... // serial path code
```